clients/tfchain-client-go/impl.go

```diff
@@ -37,145 +36,437 @@ type Versioned struct {
-// in roundrobin fashion. need to be called
-// while lock is acquired.
-func (p *mgrImpl) endpoint() string {
-	defer func() {
-		p.r = (p.r + 1) % len(p.urls)
-	}()
+func (m *manager) initializePool() {
+	log.Debug().Msg("initializing connection pool")
+	for i := 0; i < m.config.MinPoolSize; i++ {
+		select {
+		case m.checkChan <- struct{}{}:
+		default:
+		}
+	}
+}
 
-	return p.urls[p.r]
+func (m *manager) createConnection(url string) (*poolConn, error) {
+	log.Debug().Str("url", url).Msg("attempting to create a new connection")
+	if conn, meta, err := createSubstrateConn(url); err == nil {
+		log.Debug().Str("url", url).Msg("created new connection")
+		return &poolConn{
+			conn:     conn,
+			meta:     meta,
+			url:      url,
+			lastUsed: atomic.Int64{},
+			inUse:    atomic.Bool{},
+		}, nil
+	} else {
+		log.Error().Str("url", url).Err(err).Msg("failed to create connection")
+	}
+	return nil, fmt.Errorf("failed to create connection to %s", url)
 }
 
-// Substrate return a new wrapped substrate connection
-// the connection must be closed after you are done using it
-func (p *mgrImpl) Substrate() (*Substrate, error) {
-	cl, meta, err := p.Raw()
+func (m *manager) GetConnection(ctx context.Context) (*Substrate, error) {
+	log.Debug().Msg("getting a connection")
+	conn, err := m.getHealthyConn()
 	if err != nil {
-		return nil, err
-	}
-
-	return newSubstrate(cl, meta, p.put)
-}
-
-// Raw returns a RPC substrate client. plus meta. The returned connection
-// is not tracked by the pool, nor reusable. It's the caller responsibility
-// to close the connection when done
-func (p *mgrImpl) Raw() (Conn, Meta, error) {
-	// right now this pool implementation just tests the connection
-	// makes sure that it is still active, otherwise, tries again
-	// until the connection is restored.
-	// A better pool implementation can be done later were multiple connections
-	// can be handled
-	// TODO: thread safety!
-	p.m.Lock()
-	defer p.m.Unlock()
-
-	boff := backoff.WithMaxRetries(
-		backoff.NewConstantBackOff(200*time.Millisecond),
-		2*uint64(len(p.urls)),
-	)
-
-	var (
-		cl   *gsrpc.SubstrateAPI
-		meta *types.Metadata
-		err  error
-	)
-
-	err = backoff.RetryNotify(func() error {
-		endpoint := p.endpoint()
-		log.Debug().Str("url", endpoint).Msg("connecting")
-		cl, err = newSubstrateAPI(endpoint)
-		if err != nil {
-			return errors.Wrapf(err, "error connecting to substrate at '%s'", endpoint)
+		log.Error().Err(err).Msg("failed to get connection")
+		return nil, fmt.Errorf("failed to get connection: %w", err)
+	}
+	log.Debug().Str("url", conn.url).Msg("successfully obtained connection")
+	return newSubstrate(conn, m), nil
+}
+
+func (m *manager) getHealthyConn() (*poolConn, error) {
+	log.Debug().Int("pool_size", len(m.pool)).Int("aquired_count", m.aquiredConnCount()).
+		Msg("checking for healthy connections")
+
+	// Try getting existing connection first
+	if conn := m.getExistingConn(); conn != nil {
+		return conn, nil
+	}
```

Comment:
we're trying to get an existing connection here and then retrying again in the backoff retry function, should we remove this and only use the one in the retry